src/legacy/core_plugins/embeddable_api/public/containers/container.ts

```diff
@@ -0,0 +1,362 @@
+> extends Embeddable<TContainerInput, TContainerOutput>
+  implements IContainer<TContainerInput, TContainerOutput> {
+  public readonly isContainer: boolean = true;
+  protected readonly children: {
+    [key: string]: IEmbeddable<any, any> | ErrorEmbeddable;
+  } = {};
+  public readonly embeddableFactories: EmbeddableFactoryRegistry;
+
+  private subscription: Subscription;
+
+  constructor(
+    input: TContainerInput,
+    output: TContainerOutput,
+    embeddableFactories: EmbeddableFactoryRegistry,
+    parent?: Container
+  ) {
+    super(input, output, parent);
+    this.embeddableFactories = embeddableFactories;
+    this.subscription = this.getInput$().subscribe(() => this.maybeUpdateChildren());
+  }
+
+  public updateInputForChild<EEI extends EmbeddableInput = EmbeddableInput>(
+    id: string,
+    changes: Partial<EEI>
+  ) {
+    if (!this.input.panels[id]) {
+      throw new Error(
+        i18n.translate('embeddableApi.errors.paneldoesNotExist', {
+          defaultMessage: 'Panel not found.',
+        })
+      );
+    }
+    const panels = {
+      panels: {
+        ...this.input.panels,
+        [id]: {
+          ...this.input.panels[id],
+          explicitInput: {
+            ...this.input.panels[id].explicitInput,
+            ...changes,
+          },
+        },
+      },
+    };
+    this.updateInput(panels as Partial<TContainerInput>);
+  }
+
+  public reload() {
+    Object.values(this.children).forEach(child => child.reload());
+  }
+
+  public async addNewEmbeddable<
+    EEI extends EmbeddableInput = EmbeddableInput,
+    EEO extends EmbeddableOutput = EmbeddableOutput,
+    E extends IEmbeddable<EEI, EEO> = IEmbeddable<EEI, EEO>
+  >(type: string, explicitInput: Partial<EEI>): Promise<E | ErrorEmbeddable> {
+    const factory = this.embeddableFactories.getFactoryByName<EmbeddableFactory<EEI, EEO, E>>(type);
+    const panelState = this.createNewPanelState<EEI, E>(factory, explicitInput);
+
+    return new Promise(resolve => {
+      const subscription = this.getOutput$().subscribe(() => {
+        if (this.output.embeddableLoaded[panelState.embeddableId]) {
+          subscription.unsubscribe();
+          resolve(this.children[panelState.embeddableId] as E);
+        }
+      });
+
+      this.updateInput({
+        panels: {
+          ...this.input.panels,
+          [panelState.embeddableId]: panelState,
+        },
+      } as Partial<TContainerInput>);
+    });
+  }
+
+  public async addSavedObjectEmbeddable<
+    TEmbeddableInput extends EmbeddableInput = EmbeddableInput,
+    TEmbeddable extends IEmbeddable<TEmbeddableInput> = IEmbeddable<TEmbeddableInput>
+  >(type: string, savedObjectId: string): Promise<TEmbeddable | ErrorEmbeddable> {
+    const factory = this.embeddableFactories.getFactoryByName(type);
+    const panelState = this.createNewPanelState(factory);
+    panelState.savedObjectId = savedObjectId;
+
+    return new Promise(resolve => {
```

Comment:
This code seems similar to the code in `addNewEmbeddable()`.

This subscription-in-promise trick, maybe it could be extracted into a separate function:

```js
  public getChild(id) {
    return new Promise(resolve => {
      const subscription = this.getOutput$().subscribe(() => {
        if (this.output.embeddableLoaded[id]) {
          subscription.unsubscribe();
          resolve(this.children[id]);
        }
      });
    });
  }
```

Reply:
Although, you already have synchronous `getChild()`. So you basically need a wait function:

```js
  public untilEbeddableLoaded(id) {
    return new Promise(resolve => {
      const subscription = this.getOutput$().subscribe(() => {
        if (this.output.embeddableLoaded[id]) {
          subscription.unsubscribe();
          resolve();
        }
      });
    });
  }
```

and then

```ts
  public async addSavedObjectEmbeddable(type: string, savedObjectId: string): Promise<TEmbeddable | ErrorEmbeddable> {
    const factory = this.embeddableFactories.getFactoryByName(type);
    const panelState = this.createNewPanelState(factory);
    panelState.savedObjectId = savedObjectId;

    this.updateInput({
      panels: {
        ...this.input.panels,
        [panelState.embeddableId]: panelState,
      },
    } as Partial<TContainerInput>);

    const id = panelState.embeddableId;
    await this.untilEmbeddableLoaded(id);
    return this.getChild(id);
  }
```